src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -313,11 +313,11 @@ private static ImmutableArray<TextChangeRange> Merge(ImmutableArray<TextChangeRa
                             oldIndex++;
                             goto nextOldChange;
                         }
-                        else if (newChange.Span.Length == 0)
+                        else if (newChange.Span.Length <= oldChange.NewLength)
                         {
-                            // new change is just an insertion, go ahead and tack it on with old change
-                            AddRange(list, new TextChangeRange(oldChange.Span, oldChange.NewLength + newChange.NewLength));
-                            oldDelta = oldDelta - oldChange.Span.Length + oldChange.NewLength;
+                            // new change is smaller than old, go ahead and tack it on with old change
+                            AddRange(list, new TextChangeRange(oldChange.Span, oldChange.NewLength + newChange.NewLength - newChange.Span.Length));
+                            oldDelta = oldDelta - oldChange.Span.Length + oldChange.NewLength - newChange.Span.Length;
                             oldIndex++;
                             newIndex++;
```

Comment:
not certain why this is safe/ok.  if newChange and oldChange have different contents, then what's going on here?

Reply:
It actually merges the list of `TextChangeRange` objects, not `TextChange` objects.
Those don't have contents, just old length and new length, so there is no difference in contents.

The ranges are mapped to the final text later, in `GetTextChanges`.

Reply:
I don't know if this will clarify things or not, but here's the way I understand this method:

- We have an original text `t1` and call `t1.WithTextChanges(oldChanges)`. This produces a ChangedText `t2`, storing the `oldChanges` and `t1` in fields.
- When we call `t2.WithTextChanges(newChanges)` it produces a new ChangedText `t3`, storing the `newChanges` and `t2`.
- When we call `t3.GetTextChanges(t1)`, we have to rewrite the `newChanges` to make them occur w.r.t. `t1` instead of w.r.t. `t2`:
  - we have to split individual changes apart when old and new changes overlap with each other
  - we merge adjacent changes together as we go to simplify them.

In the end we get a `TextChangeRange[]` which describes the difference between the original text and the final text.